src/main/scala/SparkExample.scala

```diff
@@ -1,178 +1,336 @@
+    // TODO, Preprocess the continent column
+    // TODO: Document your function
+    // TODO: EXPORT writeCSVOutput AS A NEW FUNCTION
+    // todo: do functions documentation
+
+    /**
+     * Load CSV data with header
+     */
+    val playerData: DataFrame = readFromCSV(spark, utils.Constants.fifaPlayersDataset, hasHeader = true)
+
+    /**
+     * Load CSV data without header
+     */
+    val countriesData: DataFrame = readFromCSV(spark, utils.Constants.continentDataset,
+      hasHeader = false, columnNames = Seq("Continent", "Country"))
+
+    /**
+    * Join player dataset with countries' continent mapping
+     */
     val FifaWithContinentData = playerData.join(countriesData, playerData("Nationality") === countriesData("Country"),
         "left")
-      .drop("Country") // Drop the duplicate country column
 
-    FifaWithContinentData.show(50)
+    /**
+    * Drop the duplicate country column
+    */
+      // TODO: DOCUMENTATION BEFORE THE LINE, LEAVE A BLANK LINE BEFORE THE COMMENT
+      .drop("Country")
 
-    // Save the mixed data to CSV files partitioned by continent
-    FifaWithContinentData.write
-      .partitionBy("Continent")
-      .mode(SaveMode.Overwrite)
-      .csv("output_directory")
+    /**
+     * Caches the DataFrame 'FifaWithContinentData' into memory to optimize performance,
+     * in order to allow for faster access when used in subsequent operations.
+     * applying caching here is because the DataFrame is accessed multiple times and utilized in iterative computations.
+     */
+    FifaWithContinentData.cache()
 
-    // Load the updated salary data from the new dataset
-    val updatedSalaryData: DataFrame = spark.read
-      .option("header", "true")
-      .csv(newDatasetPath)
+    /**
+    * Save the FIFA Players data  data to CSV files partitioned by continent
+     */
+    // TODO: USE writeToCsv function
+    writeResultToCSVwithColumn(FifaWithContinentData, "output_directory", "Continent")
+
+    // todo: export new read function and use
+    // todo: pass the new dataset
+
+    val updatedSalaryData: DataFrame = readFromCSV(spark, newDatasetPath, hasHeader = true)
+    val aliasedUpdatedSalaryData = updatedSalaryData
+      .withColumnRenamed("Salary", "UpdatedSalary")
 
-    // Join the updated salary data with the existing mixed data
     // TODO: Consider having new players in the updated dataset
+    // TODO, FULL OUTER IS GREAT CHOICE, BUT USE IT RIGHT
+
+    /**
+    * Join the updated salary data with the existing entire data
+     */
+    val updatedData = FifaWithContinentData
+      .join(aliasedUpdatedSalaryData, Seq("Name", "Age", "Nationality", "Fifa Score", "Club", "Value", "Continent"),
+        "fullouter")
+
+    /**
+    * Use coalesce to merge values based on conditions
+     */
+    val mergedData = updatedData
+      .withColumn("Value", coalesce(updatedData("Value"), aliasedUpdatedSalaryData("Value")))
+      .withColumn("Salary", coalesce(FifaWithContinentData("Salary"),
```

Comment:
Great job using the coalesce here!
but for the salary we have to get the updated salary first, if null we get the existing one, else we are not updating the data.